cmd/oras/discover.go

```diff
@@ -136,19 +139,59 @@ func runDiscover(opts discoverOptions) error {
 	return nil
 }
 
-func fetchReferrers(ctx context.Context, repo *remote.Repository, desc ocispec.Descriptor, artifactType string) ([]ocispec.Descriptor, error) {
-	results := []ocispec.Descriptor{}
-	err := repo.Referrers(ctx, desc, artifactType, func(referrers []ocispec.Descriptor) error {
-		results = append(results, referrers...)
-		return nil
-	})
-	if err != nil {
-		return nil, err
+func fetchReferrers(ctx context.Context, target oras.ReadOnlyGraphTarget, desc ocispec.Descriptor, artifactType string) ([]ocispec.Descriptor, error) {
+	var results []ocispec.Descriptor
+	if repo, ok := target.(*remote.Repository); ok {
+		// get referrers directly
+		err := repo.Referrers(ctx, desc, artifactType, func(referrers []ocispec.Descriptor) error {
+			results = append(results, referrers...)
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// find matched referrers in all predecessors
+		predecessors, err := target.Predecessors(ctx, desc)
+		if err != nil {
+			return nil, err
+		}
+		for _, node := range predecessors {
+			var fetched []byte
+			if rc, err := target.Fetch(ctx, node); err != nil {
+				return nil, err
+			} else {
+				fetched, err = content.ReadAll(rc, node)
+				if err != nil {
+					return nil, err
+				}
+			}
```

Comment:
Instead of using `target.Fetch`, try [oras.FetchBytes](https://pkg.go.dev/oras.land/oras-go/v2#FetchBytes). Or make your own `FetchBytes` function, which takes a descriptor instead of a reference string.

Reply:
Will use oras.FetchBytes. Since oci target reference resolving is done in memory, it should incur minor impact on the performance.